lib/uri_checker.rb

```diff
@@ -0,0 +1,213 @@
+      elsif response.status >= 500 && response.status < 600
+        report.errors[:http_server_error] << "Received 5xx response"
+      end
+
+      report.warnings[:http_non_200] << "Page not available." unless response.status == 200
+      response
+    end
+
+    def check_get_request
+      response = make_request(:get)
+      return unless response
+
+      page = Nokogiri::HTML(response.body)
+      rating = page.css("meta[name=rating]")[0]["value"]
+      if %w(restricted mature).include?(rating)
+        report.warnings[:meta_rating] << "Page suggests it contains mature content."
+      end
+    end
+
+    def check_google_safebrowsing
+      api_key = Rails.application.secrets.google_api_key
+
+      response = Faraday.post do |req|
+        req.url "https://safebrowsing.googleapis.com/v4/threatMatches:find?key=#{api_key}"
+        req.headers["Content-Type"] = "application/json"
+        req.body = {
+          client: {
+            clientId: "gds-link-checker", clientVersion: "0.1.0"
+          },
+          threatInfo: {
+            threatTypes: %w(THREAT_TYPE_UNSPECIFIED MALWARE SOCIAL_ENGINEERING UNWANTED_SOFTWARE POTENTIALLY_HARMFUL_APPLICATION),
+            platformTypes: %w(ANY_PLATFORM),
+            threatEntryTypes: %w(URL),
+            threatEntries: [{ url: uri.to_s }]
+          }
+        }.to_json
+      end
+
+      data = JSON.parse(response.body)
+      if data.include?("matches") && data["matches"]
+        report.warnings[:google_safebrowsing] << "Google Safebrowsing has detected a threat."
+      end
+    end
+
+    def make_request(method)
+      begin
+        response = run_connection_request(method)
+
+        if response.status >= 300 && response.status < 400
+          target_uri = uri + response.headers["location"]
+          subreport = ValidUri
+            .new(redirect_history: redirect_history + [uri])
+            .call(target_uri.to_s)
+          report.merge!(subreport)
+        end
+
+        response
+      rescue Faraday::ConnectionFailed
+        report.errors[:cant_connect] << "Connection failed"
+      rescue Faraday::TimeoutError
+        report.errors[:timeout] << "Timeout Error"
+      rescue Faraday::SSLError
+        report.errors[:ssl_configuration] << "SSL Error"
+      rescue Faraday:: Error => e
+        report.errors[:unknown_http_error] << e.class.to_s
+      end
+    end
+
+    def run_connection_request(method)
+      connection.run_request(method, uri, nil, nil) do |request|
+        request.options[:timeout] = TIMEOUT
+        request.options[:open_timeout] = TIMEOUT
+      end
+    end
+
+    def connection
+      @connection ||= Faraday.new(headers: { accept_encoding: "none" }) do |faraday|
+        faraday.adapter Faraday.default_adapter
+      end
+    end
+  end
+
+  class FileChecker
+    attr_reader :report
+
+    def initialize(uri, options = {})
+      @report = Report.new
+    end
+
+    def call
+      report.errors[:local_file] << "Link is to a local file."
```

Comment:
might want to refactor this message to be a file not available on the internet (possibly your local computer)